src/main/java/com/aws/greengrass/util/RetryUtils.java

```diff
@@ -39,47 +43,77 @@ private RetryUtils() {
             "PMD.AvoidInstanceofChecksInCatchClause"})
     public static <T> T runWithRetry(RetryConfig retryConfig, CrashableSupplier<T, Exception> task,
             String taskDescription, Logger logger) throws Exception {
+        return runWithRetry(DifferentiatedRetryConfig.fromRetryConfig(retryConfig), task, taskDescription, logger);
+    }
+
+    /**
+     * Run a task with differentiated retry behaviors. Different maximum retry attempts based on different exception
+     * types. Stop the retry when interrupted.
+     * @param retryConfig     differentiated retry config
+     * @param task            task to run
+     * @param taskDescription task description
+     * @param logger          logger
+     * @param <T>             return type
+     * @return return value
+     * @throws Exception Exception
+     */
+    @SuppressWarnings({"PMD.SignatureDeclareThrowsException", "PMD.AvoidCatchingGenericException",
+            "PMD.AvoidInstanceofChecksInCatchClause"})
+    public static <T> T runWithRetry(DifferentiatedRetryConfig retryConfig, CrashableSupplier<T, Exception> task,
+                                     String taskDescription, Logger logger) throws Exception {
         long retryInterval = retryConfig.getInitialRetryInterval().toMillis();
-        int attempt = 1;
-        // if it's not the final attempt, execute and backoff on retryable exceptions
-        while (attempt < retryConfig.maxAttempt) {
+        // key = set of retryable exceptions
+        // value = current number of attempt counts
+        Map<Set<Class>, Integer> attemptMap = new HashMap<>();
+        retryConfig.getRetryMap().keySet().forEach(exceptionSet -> attemptMap.put(exceptionSet, 1));
+
+        while (true) {
```

Comment:
why is this an infinite loop now?

Reply:
added a max cap, we now compare within the loop